_algorithm_design/on_computing_next_greater_intervals_in_linear_time.md

```diff
@@ -0,0 +1,179 @@
+
+### Computing the leftBound list
+
+A naive way to compute left bound for an element $list[i]$ is to linearly search leftward in $list$ till you find the first element greater than $list[i]$. In the worst case there are $N$ states to check for every element in $list$, hence a total of $O(N^{2})$ states to search. Thus it'll take $O(N^{2})$ time to compute the list $leftBound$.
+
+However, say we already have $leftBound$ computed for the entire $list$ but we get a new element `curr` to be appended to $list$. There can be two cases here:
+- case1: $curr < list[N-1]$
+    - In this case $list[N-1]$ will trivially be the left bound of $curr$.
+- case2: $curr >= list[N-1]$
+    - Here we can use the naive way of searching linearly leftwards in $list$ till we find the first element greater than $curr$. However, one can notice from the definition of $leftBound$ that the elements in $list$ between the indices $N-1$ and $leftBound[N-1]$ are all smaller than $curr$.
+    - Thus instead of searching linearly leftwards, we could skip the elements in between and directly check $list[leftBound[N-1]]$, then $list[leftBound[leftBound[N-1]]]$ and so on until we find the element greater than $curr$.
+    - The set of indices $list[N-1]$, $list[leftBound[N-1]]$, $list[leftBound[leftBound[N-1]]]$ and so on form a topology of ancestors dictated strictly by ordering with respect to the $>$ operator. At this point we can safely conclude that computation of left bound for a new element depends strictly on the topological ancestors of the last element in the $list$.
+
+Hence we have pruned the search space but by how much? We will discuss the answer to this shortly.
+
+#### The monotonic stack
+
+There are two noteworthy observations to make considering that we need only maintain the topological ancestors of the last element in the $list$ to compute the left bound of $curr$:
+- case1 tells us that $curr$ can simply be added to the front of the topology without disturbing the rest of its structure. case2 tells us that once we find the left bound of $curr$, we don't need to maintain any elements between $curr$ and its left bound. If you stare at the screen long enough, you'll notice that case1 and case2 together manifest a `last-in-first-out` behaviour in the topology of ancestors. Hence a stack can be used to represent the topological ancestors of the last element in $list$.
+- case2 also suggests that the topology obeys the following invariant $ancestor(item) > item \, \forall item \in topology$. The invariant is equivalent to monotonicity in the stack.
+
+Thus monotonic stack gets its name from the two observations made above. So far so good! Now we are in a position to write down the semantics of the proposed data structure. I am (again) choosing Pythonic pseudo code to write the semantics.
+
+{% highlight python %}
+class MonotonicStack(Stack):
+
+    @abstractmethod
+    def is_empty():
```

Comment:
follow camelCase maybe

Reply:
PEP dictates that method names shouldn't be camelCased.